### PR TITLE
Support deeper trees by using iteration instead of recursion

### PR DESCRIPTION
**Pull request by lopuhin**
in master we crash on deep trees with a recursion error

### CODE REVIEW

**Review by BurnzZ — APPROVED** [review]
LGTM @lopuhin !

File: tests/test_html_text.py [hunk 1]
```diff
@@ -214,3 +214,23 @@ def test_webpages(page, extracted):
 
     tree = cleaner.clean_html(parse_html(html))
     assert etree_to_text(tree) == expected
+
+
+def test_deep_html():
+    """ Make sure we don't crash due to recursion limit.
+    """
+    # Build a deep tree manually as default parser would only allow
+    # for 255 depth, but deeper trees are possible with other parsers
```
Comment (kmike) [inline tests/test_html_text.py:223]: We might want to copy https://github.com/scrapy/parsel/blob/740734252efd246679e5c6ee4028a47d5851712d/parsel/selector.py#L87 to html-text (not in this PR)

File: html_text/html_text.py [hunk 2]
```diff
@@ -132,17 +132,17 @@ def add_text(text_content, context):
...
-        add_text(tree.text, context)
-        for child in tree:
-            traverse_text_fragments(child, context)
-        add_newlines(tree.tag, context)
-        if handle_tail:
-            add_text(tree.tail, context)
-
-    traverse_text_fragments(tree, context=Context(), handle_tail=False)
+    # Extract text from the ``tree``: fill ``chunks`` variable
+    context = Context()
```
Comment (kmike) [inline html_text/html_text.py:136]: It seems we can drop Context, and just use nonlocal. Not required to get this PR merged, of course :)

**Review by kmike — APPROVED** [review]
Nice!

File: tests/test_html_text.py (hunk 1 above)
Comment (lopuhin) [reply tests/test_html_text.py:223]: Good point @kmike , I created https://github.com/zytedata/html-text/issues/9

File: html_text/html_text.py (hunk 2 above)
Comment (lopuhin) [reply html_text/html_text.py:136]: Thanks @kmike implemented in https://github.com/zytedata/html-text/pull/10

**Comment by lopuhin** [comment]
Thanks for review @kmike and @BurnzZ 👍

**Pull request merged**